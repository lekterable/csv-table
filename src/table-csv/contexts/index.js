import { createContext } from 'react'

export const ModeContext = createContext({ isEditable: false })
export const ContentContext = createContext([
  { content: [], uniqueKeys: [] },
  () => {}
])
