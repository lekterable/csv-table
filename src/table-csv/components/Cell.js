import React, { useContext, useState } from 'react'
import { ContentContext } from '../contexts'

export default ({ header, value, rowIndex, colIndex }) => {
  const [, dispatch] = useContext(ContentContext)
  const [content, setContent] = useState(value)
  const [editMode, setEditMode] = useState(false)
  const toggleMode = () => setEditMode(mode => !mode)
  const handleFocus = ref => ref && ref.focus()
  const handleChange = e => setContent(e.target.value)
  const submitContent = () => {
    dispatch({ type: 'UPDATE_CONTENT', content, rowIndex, colIndex })
    toggleMode()
  }
  const handleKeyDown = e => {
    if (e.key === 'Enter') return submitContent()
    else if (e.key === 'Escape') {
      toggleMode()
      return setContent(value)
    }
  }

  return header ? (
    <th>
      {editMode ? (
        <input
          type="text"
          value={content}
          onBlur={submitContent}
          onChange={handleChange}
          onKeyDown={handleKeyDown}
          ref={handleFocus}
        />
      ) : (
        <div onClick={toggleMode}>{content}</div>
      )}
    </th>
  ) : (
    <td>
      {editMode ? (
        <input
          type="text"
          value={content}
          onBlur={submitContent}
          onChange={handleChange}
          onKeyDown={handleKeyDown}
          ref={handleFocus}
        />
      ) : (
        <div onClick={toggleMode}>{content}</div>
      )}
    </td>
  )
}
