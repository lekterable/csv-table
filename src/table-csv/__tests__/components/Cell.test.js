import React from 'react'
import renderer from 'react-test-renderer'
import { cleanup, fireEvent, render } from 'react-testing-library'
import Cell from '../../components/Cell'
import { ContentContext, ModeContext } from '../../contexts'

describe('<Cell />', () => {
  const withTable = component => (
    <table>
      <tbody>
        <tr>{component}</tr>
      </tbody>
    </table>
  )
  const withContext = (component, mode, content) => (
    <ModeContext.Provider value={mode}>
      <ContentContext.Provider value={content}>
        {component}
      </ContentContext.Provider>
    </ModeContext.Provider>
  )

  afterEach(cleanup)

  it('should match snapshot', () => {
    const component = renderer.create(<Cell />).toJSON()
    expect(component).toMatchSnapshot()
  })

  it('should render cell', () => {
    const { queryByTestId } = render(withTable(<Cell />))
    const headerCell = queryByTestId('table-header-cell')
    const cell = queryByTestId('table-cell')

    expect(headerCell).toBeNull()
    expect(cell).not.toBeNull()
  })

  it('should render header cell if `header` prop', () => {
    const { queryByTestId } = render(withTable(<Cell header />))
    const headerCell = queryByTestId('table-header-cell')
    const cell = queryByTestId('table-cell')

    expect(headerCell).not.toBeNull()
    expect(cell).toBeNull()
  })

  it('should not render input if not editable', () => {
    const dispatchSpy = jest.fn()

    const { getByTestId, queryByTestId } = render(
      withContext(
        withTable(<Cell value="Just a Cell" colIndex={2} rowIndex={3} />),
        { isEditable: false },
        [{ content: [], uniqueKeys: [] }, dispatchSpy]
      )
    )

    fireEvent.click(getByTestId('editable-field-display'))
    const input = queryByTestId('editable-field-input')

    expect(input).toBeNull()
  })

  it('should update the global content on submit', () => {
    const dispatchSpy = jest.fn()

    const { getByTestId } = render(
      withContext(
        withTable(<Cell value="Just a Cell" colIndex={2} rowIndex={3} />),
        { isEditable: true },
        [{ content: [], uniqueKeys: [] }, dispatchSpy]
      )
    )

    fireEvent.click(getByTestId('editable-field-display'))
    const input = getByTestId('editable-field-input')
    fireEvent.change(input, { target: { value: "I'm a updated Cell" } })
    fireEvent.keyDown(input, { key: 'Enter' })
    const display = getByTestId('editable-field-display')

    expect(display.textContent).toBe("I'm a updated Cell")
    expect(dispatchSpy).toBeCalled()
    expect(dispatchSpy).toBeCalledWith({
      colIndex: 2,
      content: "I'm a updated Cell",
      rowIndex: 3,
      type: 'UPDATE_CONTENT'
    })
  })

  it('should discard the changes on escape', () => {
    const dispatchSpy = jest.fn()

    const { getByTestId } = render(
      withContext(
        withTable(<Cell value="Just a Cell" colIndex={2} rowIndex={3} />),
        { isEditable: true },
        [{ content: [], uniqueKeys: [] }, dispatchSpy]
      )
    )

    fireEvent.click(getByTestId('editable-field-display'))
    const input = getByTestId('editable-field-input')
    fireEvent.change(input, { target: { value: "I'm a updated Cell" } })
    fireEvent.keyDown(input, { key: 'Space' })
    fireEvent.keyDown(input, { key: 'Escape' })
    const display = getByTestId('editable-field-display')

    expect(display.textContent).toBe('Just a Cell')
    expect(dispatchSpy).not.toBeCalled()
  })
})
