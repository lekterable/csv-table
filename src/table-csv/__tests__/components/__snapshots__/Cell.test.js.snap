// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cell /> should match snapshot 1`] = `
<td
  className="table__cell"
  data-testid="table-cell"
>
  <div
    className="editable-field"
    data-testid="editable-field-display"
    onClick={[Function]}
  >
    
  </div>
</td>
`;
